### PR TITLE
Metadata proxy

### PR DESCRIPTION
**Pull request by Q-Lee**
Adding a container to proxy the gce metadata service.

### CODE REVIEW

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: Please don't use this image. I can't tell what it's sources are, and it hasn't been updated in 2 years. This should use either gcr.io/google-containers/debian-base-amd64 or gcr.io/google-containers/debian-iptables

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: nit: maintainer tags are deprecated

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: I think it's weird to run sudo in a container... it should probably just run as root. If you're trying to avoid running nginx as root, I would just switch to a different user for that command.

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: nit: I think `COPY` preserves permissions, so you could just add this permission to the file checked into git and skip this step.

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: nit: prefer `ENTRYPOINT`

File: metadata-proxy/Makefile
Comment (timstclair) [inline metadata-proxy/Makefile]: I think you might need license headers on each file.

File: metadata-proxy/Makefile
Comment (timstclair) [inline metadata-proxy/Makefile]: delete this

File: metadata-proxy/Makefile
Comment (timstclair) [inline metadata-proxy/Makefile]: nit: do you want to start with a pre-release version? 0.1?

File: metadata-proxy/nginx.conf
Comment (timstclair) [inline metadata-proxy/nginx.conf]: Can we use a whitelist instead of a blacklist approach? My concern is that, for example, when new API versions (or new APIs) come out, suddenly every node has access.

File: metadata-proxy/start-proxy.sh
Comment (timstclair) [inline metadata-proxy/start-proxy.sh]: I don't think you need this with the `daemon off` setting?

File: metadata-proxy/nginx.conf
Comment (timstclair) [inline metadata-proxy/nginx.conf]: how did you pick :2020?

**Review by timstclair — CHANGES_REQUESTED** [review]
Can you add the daemonset manifest too? Even if it's not the final version, I think including the manifest provides a good example of how some of the options work (e.g. I assume this needs host networking)

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Ok, I changed it to be a label.

**Comment by Q-Lee** [comment]
@timstclair - the daemonset will live in the main repo as an addon. I haven't sent it out yet, because I don't want to add anything to "google-containers" until it's been reviewed and checked in.

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Ok, I'm following fluentd's lead and using "gcr.io/google-containers/debian-base-amd64:0.1".

File: metadata-proxy/Makefile
Comment (Q-Lee) [inline metadata-proxy/Makefile]: Added.

File: metadata-proxy/Makefile
Comment (Q-Lee) [inline metadata-proxy/Makefile]: Done.

File: metadata-proxy/nginx.conf
Comment (Q-Lee) [inline metadata-proxy/nginx.conf]: Nothing was using it, and dynamic allocation would be considerably more difficult. Do you have another proposal?

Comment (timstclair) [inline metadata-proxy/nginx.conf]: No, just curious. Can you use variables in the nginx configuration? It's not too bad since this is optional, but what if a customer has something running on that port? I would be nice to provide a way to override it.

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Ok, I removed sudo. The nginx master process starts as root, and then forks www-data worker processes.

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Ok

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Ok

File: metadata-proxy/nginx.conf
Comment (Q-Lee) [inline metadata-proxy/nginx.conf]: I don't think nginx configs support variables. In fact, I think their developers staunchly oppose that feature. IIRC, they recommend sed.

I could move it to a privileged port, and then only privileged pods w/host networking could conflict, instead of just host networking. I'll also drop this to v0.1, in case we decide the value should be configurable.

File: metadata-proxy/nginx.conf
Comment (Q-Lee) [inline metadata-proxy/nginx.conf]: That's fair. I've switched to a whitelist. It makes a few things (e.g., REST discovery) weird, so I've aimed for clarity over brevity.

BTW, this config is overwritten by a configmap in the daemonset. I'm included this for the same reason as the "ENTRYPOINT": sane defaults. This allows users with sufficient permissions to overwrite the behavior.

File: metadata-proxy/start-proxy.sh
Comment (Q-Lee) [inline metadata-proxy/start-proxy.sh]: I don't.

File: metadata-proxy/Makefile
Comment (Q-Lee) [inline metadata-proxy/Makefile]: Ok

**Comment by Q-Lee** [comment]
@timstclair - does this look better?

Take a look at the configmap and daemonset here:
https://github.com/Q-Lee/kubernetes/tree/mds/cluster/addons/metadata-proxy

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: Use `clean-install iptables nginx`

File: metadata-proxy/Dockerfile
Comment (timstclair) [inline metadata-proxy/Dockerfile]: I thought you were removing sudo?

File: metadata-proxy/nginx.conf
Comment (timstclair) [inline metadata-proxy/nginx.conf]: So I'll let you decide if it's worth it in this PR or not, but this is what you could do here:

1. Add variables for `PORT` and `METADATA_IP` (e.g. `__PORT__` and `__METADATA_IP__`)
2. Update the start-proxy.sh script to replace those variables in the config with the value from an environment variable, using sed, e.g. `sed -i s/__PORT__/${PORT:-988}/g /etc/nginx/nginx.conf`
3. Populate the env vars from the pod spec

File: metadata-proxy/start-proxy.sh
Comment (timstclair) [inline metadata-proxy/start-proxy.sh]: delete this?

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: I sure did.

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: What's "clean-install"?

Comment (timstclair) [inline metadata-proxy/Dockerfile]: A script that encapsulates `apt-get update && apt-get install` and also does cleanup: https://github.com/kubernetes/kubernetes/blob/master/build/debian-base/clean-install

File: metadata-proxy/nginx.conf
Comment (Q-Lee) [inline metadata-proxy/nginx.conf]: Even if make the port configurable, the addon is still a static file on the master. I think that choosing an unassigned privileged port does more to prevent conflicts than hiding an easy-to-configure setting behind a hard-to-impossible-to-configure implementation.

We can change this if it becomes a problem, but I don't see it being useful until we decide on, implement, and deploy a serious overhaul to the addon manager.

File: metadata-proxy/start-proxy.sh
Comment (Q-Lee) [inline metadata-proxy/start-proxy.sh]: Whoops.

File: metadata-proxy/Dockerfile
Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: The "apt-utils" package must be installed first, and "apt-get install" doesn't guarantee the order of installation.

I imagine the next two commands (clean and rm -rf) are simply to shrink the image size?

Comment (Q-Lee) [inline metadata-proxy/Dockerfile]: Interesting.

no cleanup:        fb786fa5fc50        112 MB
"clean-install":        730191ef3d6f        61.94 MB
"apt-get clean -y":        7f7e7eac1fb7        112 MB
gcr.io/google-containers/debian-base-amd64:        61a750f1056b        40.64 MB

The "rm -rf" seems to be removing about 50MB of junk. I'm convinced.

**Comment by timstclair** [comment]
lgtm. Please squash & rebase.

**Comment by jbeda** [comment]
This image should probably be named `gce-metadata-proxy` as it is GCE specific.  Might be a good idea to fix that up before the close of 1.7.

**Pull request merged**
